leakcanary-deobfuscation-gradle-plugin/src/main/java/com/squareup/leakcanary/deobfuscation/LeakcanaryLeakDeobfuscationPlugin.kt

```diff
@@ -0,0 +1,111 @@
+package com.squareup.leakcanary.deobfuscation
+
+import com.android.build.gradle.AppExtension
+import com.android.build.gradle.BaseExtension
+import com.android.build.gradle.LibraryExtension
+import com.android.build.gradle.api.ApplicationVariant
+import com.android.build.gradle.api.BaseVariant
+import com.android.build.gradle.api.LibraryVariant
+import org.gradle.api.DefaultTask
+import org.gradle.api.DomainObjectSet
+import org.gradle.api.GradleException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.UnknownTaskException
+import org.gradle.api.tasks.TaskProvider
+import java.io.File
+
+class LeakcanaryLeakDeobfuscationPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    val androidPluginHandler = { _: Plugin<*> ->
+      val extension = project.extensions.getByType(BaseExtension::class.java)
+      val variants: DomainObjectSet<BaseVariant> = when (extension) {
+        is AppExtension -> {
+          extension.applicationVariants as DomainObjectSet<BaseVariant>
+        }
+        is LibraryExtension -> {
+          extension.libraryVariants as DomainObjectSet<BaseVariant>
+        }
+        else -> throw GradleException(
+            "This plugin can only be applied in Android application or library modules."
+        )
+      }
+      val leakCanaryExtension = project.extensions.create(
+          "leakCanary",
+          LeakCanaryDeobfuscationExtension::class.java
+      )
+
+      variants.all { variant ->
+        val obfuscatedVariant =
+          leakCanaryExtension.filterObfuscatedVariants?.invoke(variant) ?: false
+        if (obfuscatedVariant || variant.buildType.isMinifyEnabled) {
+          setupPlugin(project, variant)
+        }
+      }
+    }
+
+    when {
```

Comment:
Why don't we make that more simple: instead of having this structure of a plugin handler and use `withId`, why not simply start in `apply` by assuming an android plugin is there. With a few more internal methods, the apply method would look like this:

```kotlin
override fun apply(project: Project) {
  val variants = findAndroidVariants(project)
  val leakCanaryExtension = createLeakCanaryExtension(project)
  variants.all { variant ->
        val obfuscatedVariant =
          leakCanaryExtension.filterObfuscatedVariants?.invoke(variant) ?: false
        if (obfuscatedVariant) {
          setupTasks(project, variant)
        }
      }
}

Reply:
Makes sense, will try to do it this way. I checked that SqlDelight plugin uses different approach:
https://github.com/cashapp/sqldelight/commit/8cf2bf288f31fa94ab09318b6080aa987f10978b

they use `plugins.withId` + `project.afterEvaluate` + a boolean flag to make sure that their tasks are set up after android plugin finishes setting up its variants.

Reply:
then the plugin would be less sensitive to ordering.. I understand, but I find the code harder to read. I keep it in mind though. If you want to follow this track, up to you, we can make the code more concise. Not sure we need the boolean check. Do we ?

Reply:
I agree that it's better to keep it simpler and more concise. I'll remove `withId`.